<a href="https://colab.research.google.com/github/Priyankamishra2098/Global-terrorism-datset-EDA/blob/main/EDA_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **1.Importing Libraries and Dataset**

### Importing Libraries

In [1]:
#importing libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
from wordcloud import WordCloud
import warnings
warnings.filterwarnings('ignore')
print("Libraries imported sucessfully")


Libraries imported sucessfully


### Loading Dataset

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
df = "/content/drive/MyDrive/data/Global Terrorism Data.csv"
print("Dataset loaded sucessfully")

Dataset loaded sucessfully


### Reading of Dataset

In [4]:
data_ter = pd.read_csv(df,encoding = 'ISO-8859-1') 
print('Dataset Read Successfully')

Dataset Read Successfully


###Dimensions of Dataset

In [5]:
data_ter.shape

(181691, 135)

### Dataset First View

In [6]:
data_ter.head(3)

eventid  iyear  imonth  iday approxdate  extended resolution  country  \
0  197000000001   1970       7     2        NaN         0        NaN       58   
1  197000000002   1970       0     0        NaN         0        NaN      130   
2  197001000001   1970       1     0        NaN         0        NaN      160   

          country_txt  region  ... addnotes scite1 scite2  scite3  dbsource  \
0  Dominican Republic       2  ...      NaN    NaN    NaN     NaN      PGIS   
1              Mexico       1  ...      NaN    NaN    NaN     NaN      PGIS   
2         Philippines       5  ...      NaN    NaN    NaN     NaN      PGIS   

   INT_LOG  INT_IDEO INT_MISC INT_ANY  related  
0        0         0        0       0      NaN  
1        0         1        1       1      NaN  
2       -9        -9        1       1      NaN  

[3 rows x 135 columns]

In [7]:
data_ter.tail()

eventid  iyear  imonth  iday approxdate  extended resolution  \
181686  201712310022   2017      12    31        NaN         0        NaN   
181687  201712310029   2017      12    31        NaN         0        NaN   
181688  201712310030   2017      12    31        NaN         0        NaN   
181689  201712310031   2017      12    31        NaN         0        NaN   
181690  201712310032   2017      12    31        NaN         0        NaN   

        country  country_txt  region  ... addnotes  \
181686      182      Somalia      11  ...      NaN   
181687      200        Syria      10  ...      NaN   
181688      160  Philippines       5  ...      NaN   
181689       92        India       6  ...      NaN   
181690      160  Philippines       5  ...      NaN   

                                                   scite1  \
181686  "Somalia: Al-Shabaab Militants Attack Army Che...   
181687  "Putin's 'victory' in Syria has turned into a ...   
181688  "Maguindanao clashes trap tribe members," Phil...   
181689  "Trader escapes grenade attack in Imphal," Bus...   
181690  "Security tightened in Cotabato following IED ...   

                                                   scite2  \
181686  "Highlights: Somalia Daily Media Highlights 2 ...   
181687  "Two Russian soldiers killed at Hmeymim base i...   
181688                                                NaN   
181689                                                NaN   
181690  "Security tightened in Cotabato City," Manila ...   

                                                   scite3  \
181686  "Highlights: Somalia Daily Media Highlights 1 ...   
181687  "Two Russian servicemen killed in Syria mortar...   
181688                                                NaN   
181689                                                NaN   
181690                                                NaN   

                        dbsource  INT_LOG  INT_IDEO INT_MISC INT_ANY  related  
181686  START Primary Collection        0         0        0       0      NaN  
181687  START Primary Collection       -9        -9        1       1      NaN  
181688  START Primary Collection        0         0        0       0      NaN  
181689  START Primary Collection       -9        -9        0      -9      NaN  
181690  START Primary Collection       -9        -9        0      -9      NaN  

[5 rows x 135 columns]

## **2.Understanding the Variables**

###Dataset Information

In [8]:
data_ter.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 181691 entries, 0 to 181690
Columns: 135 entries, eventid to related
dtypes: float64(55), int64(22), object(58)
memory usage: 187.1+ MB


In [9]:
data_ter.columns

Index(['eventid', 'iyear', 'imonth', 'iday', 'approxdate', 'extended',
       'resolution', 'country', 'country_txt', 'region',
       ...
       'addnotes', 'scite1', 'scite2', 'scite3', 'dbsource', 'INT_LOG',
       'INT_IDEO', 'INT_MISC', 'INT_ANY', 'related'],
      dtype='object', length=135)

### Number of Duplicate Values in the Dataset

In [10]:
# Duplicate Values/Duplicate Values Count
data_ter.duplicated().sum()

0

### Null Values in the Dataset

In [11]:
# Missing Values/Null Values Count
pd.DataFrame(data_ter.isna().sum()).T

eventid  iyear  imonth  iday  approxdate  extended  resolution  country  \
0        0      0       0     0      172452         0      179471        0   

   country_txt  region  ...  addnotes  scite1  scite2  scite3  dbsource  \
0            0       0  ...    153402   66191  104758  138175         0   

   INT_LOG  INT_IDEO  INT_MISC  INT_ANY  related  
0        0         0         0        0   156653  

[1 rows x 135 columns]

###Total Number of Null Values

In [12]:
# Counting Total Number of Null Values
print('Total number of null values = ',data_ter.isnull().sum().sum())

Total number of null values =  13853997


###Checking Columns with more then 50% null values

In [13]:
# Calculating the percentage of null values in each column
null_percentages = data_ter.isnull().sum() / len(data_ter)

# Filtering columns with more than 50% null values
null_columns = null_percentages[null_percentages > 0.5].index.tolist()

# Printing the columns with more than 50% null values
print(null_columns)


['approxdate', 'resolution', 'location', 'alternative', 'alternative_txt', 'attacktype2', 'attacktype2_txt', 'attacktype3', 'attacktype3_txt', 'targtype2', 'targtype2_txt', 'targsubtype2', 'targsubtype2_txt', 'corp2', 'target2', 'natlty2', 'natlty2_txt', 'targtype3', 'targtype3_txt', 'targsubtype3', 'targsubtype3_txt', 'corp3', 'target3', 'natlty3', 'natlty3_txt', 'gsubname', 'gname2', 'gsubname2', 'gname3', 'gsubname3', 'motive', 'guncertain2', 'guncertain3', 'claimmode', 'claimmode_txt', 'claim2', 'claimmode2', 'claimmode2_txt', 'claim3', 'claimmode3', 'claimmode3_txt', 'compclaim', 'weaptype2', 'weaptype2_txt', 'weapsubtype2', 'weapsubtype2_txt', 'weaptype3', 'weaptype3_txt', 'weapsubtype3', 'weapsubtype3_txt', 'weaptype4', 'weaptype4_txt', 'weapsubtype4', 'weapsubtype4_txt', 'propextent', 'propextent_txt', 'propvalue', 'propcomment', 'nhostkid', 'nhostkidus', 'nhours', 'ndays', 'divert', 'kidhijcountry', 'ransom', 'ransomamt', 'ransomamtus', 'ransompaid', 'ransompaidus', 'ransomnot

###Correlation Matrix

In [14]:
# calculating the correlation matrix
corr_matrix = data_ter.corr()

# printing the correlation matrix
print(corr_matrix)

            eventid     iyear    imonth      iday  extended   country  \
eventid    1.000000  0.999996  0.002706  0.018336  0.091761 -0.135039   
iyear      0.999996  1.000000  0.000139  0.018254  0.091754 -0.135023   
imonth     0.002706  0.000139  1.000000  0.005497 -0.000468 -0.006305   
iday       0.018336  0.018254  0.005497  1.000000 -0.004700  0.003468   
extended   0.091761  0.091754 -0.000468 -0.004700  1.000000 -0.020466   
...             ...       ...       ...       ...       ...       ...   
nreleased -0.181612 -0.181556 -0.011535  0.001765 -0.192155 -0.044331   
INT_LOG   -0.143600 -0.143601 -0.002302 -0.001540  0.071768  0.069904   
INT_IDEO  -0.133252 -0.133253 -0.002034 -0.001621  0.075147  0.067564   
INT_MISC  -0.077852 -0.077847 -0.002554 -0.002027  0.027335  0.207281   
INT_ANY   -0.175605 -0.175596 -0.006336 -0.001199  0.080767  0.153118   

             region  latitude  longitude  specificity  ...  ransomamt  \
eventid    0.401371  0.166886   0.003907     0.030

###**3.Pre-Processing and Cleaning of Data**

###Dropping any column in the DataFrame df that contains all missing values (i.e., NaN).

In [15]:
data_ter.dropna(axis=1,how='all')
data_ter.head(2)

eventid  iyear  imonth  iday approxdate  extended resolution  country  \
0  197000000001   1970       7     2        NaN         0        NaN       58   
1  197000000002   1970       0     0        NaN         0        NaN      130   

          country_txt  region  ... addnotes scite1 scite2  scite3  dbsource  \
0  Dominican Republic       2  ...      NaN    NaN    NaN     NaN      PGIS   
1              Mexico       1  ...      NaN    NaN    NaN     NaN      PGIS   

   INT_LOG  INT_IDEO INT_MISC INT_ANY  related  
0        0         0        0       0      NaN  
1        0         1        1       1      NaN  

[2 rows x 135 columns]

In [16]:
#replacing Na values of every column with dtype as float64 by  mean 
#list_num_col = data_ter.select_dtypes(include = ["float64"]).columns.to_list()
#for i in list_num_col:
 # data_ter.loc[:, i] = data_ter[i].fillna(data_ter[i].mean())


In [17]:
#replacing Na values of every column with dtype as object by mode
#list_str_col = data_ter.select_dtypes(include = ["object"]).columns.to_list()
#for i in list_num_col:
  #data_ter.loc[:, i] = data_ter[i].fillna(data_ter[i].mode()[0])


###Renaming and Selecting Required Columns

In [18]:
data_ter.rename(columns={'iyear':'Year','imonth':'Month','iday':'Day','eventid':'Eventid',
                   'approxdate':'Date','region_txt':'Region',
                   'country_txt':'Country','attacktype1_txt':'Attack_Type',
                   'target1':'Target','nkill':'Killed','nwound':'Wounded',
                   'gname':'Group','targtype1_txt':'Target_Type',
                   'weaptype1_txt':'Weapon_Type','provstate':'State','motive':'Motive'},inplace=True)

In [19]:
df = data_ter[['Year','Month','Day','Country','State','Region','city','Motive',
           'latitude','longitude','Attack_Type','Killed','Wounded','Target',
           'summary','Group','Target_Type','Weapon_Type']]
df['Casualities'] = df['Killed']+df['Wounded']


###Extracting New Data

In [20]:
df.head(2)

Year  Month  Day             Country    State                       Region  \
0  1970      7    2  Dominican Republic      NaN  Central America & Caribbean   
1  1970      0    0              Mexico  Federal                North America   

            city Motive   latitude  longitude                  Attack_Type  \
0  Santo Domingo    NaN  18.456792 -69.951164                Assassination   
1    Mexico city    NaN  19.371887 -99.086624  Hostage Taking (Kidnapping)   

   Killed  Wounded                   Target summary  \
0     1.0      0.0             Julio Guzman     NaN   
1     0.0      0.0  Nadine Chaval, daughter     NaN   

                                Group                  Target_Type  \
0                              MANO-D  Private Citizens & Property   
1  23rd of September Communist League      Government (Diplomatic)   

  Weapon_Type  Casualities  
0     Unknown          1.0  
1     Unknown          0.0

###Shape of Processed Data

In [21]:
df.shape

(181691, 19)

### Modifications Done and Insights Gained

This Global Terrorism Dataset (GTD) is a widely used dataset on terrorism worldwide, covering incidents from 1970 to 2017. The GTD includes a wide range of variables for each incident.Thus, the dataset is extremely large, consisting of 181691 rows and 135 columns with data types : float64(55), int64(22), object(58).

Upon checking for duplicate values i got to know that there are no duplicate values in the dataset.Checking for null values in a dataset is also important step in data cleaning and preparation, as null values can have a significant impact on data analysis and modeling. So to ensure that data is accurate and reliable, and that it can be used effectively I identified the null values with total count of 13853997, and then handled by Dropping any column in the DataFrame df that contains all missing values (i.e., NaN).



